components/public-api-server/pkg/apiv1/workspace.go

```diff
@@ -119,3 +124,148 @@ func bearerTokenFromContext(ctx context.Context) (string, error) {
 	token := values[0]
 	return token, nil
 }
+
+func (w *WorkspaceService) ListWorkspaces(ctx context.Context, req *v1.ListWorkspacesRequest) (*v1.ListWorkspacesResponse, error) {
+	logger := ctxlogrus.Extract(ctx)
+	token, err := bearerTokenFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := w.connectionPool.Get(ctx, token)
+	if err != nil {
+		logger.WithError(err).Error("Failed to get connection to server.")
+		return nil, status.Error(codes.Internal, "failed to establish connection to downstream services")
+	}
+
+	serverResp, err := server.GetWorkspaces(ctx, &protocol.GetWorkspacesOptions{
+		Limit: float64(getLimitFromPagination(req.Pagination)),
+	})
+	if err != nil {
+		return nil, proxy.ConvertError(err)
+	}
+
+	res := make([]*v1.ListWorkspacesResponse_WorkspaceAndInstance, 0, len(serverResp))
+	for _, ws := range serverResp {
+		workspaceAndInstance, err := convertWorkspaceInfo(ws)
+		if err != nil {
+			// TODO(cw): not sure this is a good idea - a single invalid workspace can break the entire response
```

Comment:
I think this is better than silently skipping the entry. I'd keep it like this, except I'd wrap the error in more context, such as `fmt.Errorf("failed to convert workspace instance ID: %s into response: %w", instanceID, err)`

Reply:
Didn't find anything in 43min
<img width="648" alt="image" src="https://user-images.githubusercontent.com/3210701/178528382-eb59c88c-7273-4ce6-b2cf-90be5a889b22.png">